File: 20-actors/mitooshi/methods/test_persist.py
#!/usr/bin/env python3
"""Tests for the mitooshi append-only chokepoint-intel persistence (methods/persist.py).

    PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python3 -m pytest test_persist.py
    python3 test_persist.py

The load-bearing property under test: the trail is APPEND-ONLY (非終末論). A re-run is
idempotent; a new snapshot is additive; an existing observation is NEVER removed or
mutated; the emitted EDN round-trips through the same reader a live ingest would use.
"""
from __future__ import annotations

import pathlib
import sys
import tempfile

try:
    from bridge import bridge
    from analyze import load_edn
    from persist import (append_obs, emit_trail_edn, load_trail, merge_series, persist)
except ImportError:
    from mitooshi.methods.bridge import bridge  # type: ignore
    from mitooshi.methods.analyze import load_edn  # type: ignore
    from mitooshi.methods.persist import (  # type: ignore
        append_obs, emit_trail_edn, load_trail, merge_series, persist)

_BRIDGE = pathlib.Path(__file__).resolve().parent.parent / "data" / "bridge"


def _by_actor():
    return {
        "watari": load_edn(_BRIDGE / "watari-sample.edn"),
        "watatsuna": load_edn(_BRIDGE / "watatsuna-sample.edn"),
    }


def test_append_to_empty_trail_adds_all():
    b = bridge(_by_actor(), observed_at=1)
    merged, added, dup = append_obs([], b["obs"])
    assert added == len(b["obs"]) and dup == 0
    assert len(merged) == len(b["obs"])


def test_reappend_same_snapshot_is_idempotent():
    b = bridge(_by_actor(), observed_at=1)
    merged, _, _ = append_obs([], b["obs"])
    merged2, added2, dup2 = append_obs(merged, b["obs"])
    assert added2 == 0 and dup2 == len(b["obs"])      # nothing new, all duplicates
    assert len(merged2) == len(merged)                 # trail unchanged


def test_new_snapshot_is_additive_never_removes():
    b1 = bridge(_by_actor(), observed_at=1)
    b2 = bridge(_by_actor(), observed_at=2)
    merged, _, _ = append_obs([], b1["obs"])
    before_ids = {o[":obs/id"] for o in merged}
    merged2, added2, dup2 = append_obs(merged, b2["obs"])
    after_ids = {o[":obs/id"] for o in merged2}
    assert before_ids <= after_ids                     # 非終末論: never drops an obs
    assert added2 == len(b2["obs"]) and dup2 == 0      # different ts → all new


def test_existing_obs_values_are_not_mutated():
    b1 = bridge(_by_actor(), observed_at=1)
    merged, _, _ = append_obs([], b1["obs"])
    snapshot = {o[":obs/id"]: o[":obs/value"] for o in merged}
    # a later (hypothetically revised) snapshot at the same ts must NOT overwrite
    merged2, added, dup = append_obs(merged, b1["obs"])
    after = {o[":obs/id"]: o[":obs/value"] for o in merged2}
    assert after == snapshot and added == 0


def test_merge_series_is_union_first_wins():
    b = bridge(_by_actor(), observed_at=1)
    merged = merge_series({}, b["series"])
    assert set(merged) == set(b["series"])
    # re-merging keeps the first definition (stable identity)
    again = merge_series(merged, b["series"])
    assert again == merged


def test_emit_round_trips_through_reader():
    b = bridge(_by_actor(), observed_at=1)
    merged_obs, _, _ = append_obs([], b["obs"])
    edn = emit_trail_edn(b["series"], merged_obs)
    recs = load_edn_from_str(edn)
    obs = [r for r in recs if ":obs/id" in r]
    series = [r for r in recs if ":series/id" in r]
    assert len(obs) == len(merged_obs) and len(series) == len(b["series"])
    # values survive the round-trip
    vals = sorted(o[":obs/value"] for o in obs)
    assert vals == sorted(o[":obs/value"] for o in merged_obs)


def test_persist_to_disk_two_snapshots_accumulate():
    with tempfile.TemporaryDirectory() as d:
        trail = pathlib.Path(d) / "chokepoint-trail.kotoba.edn"
        s1 = persist(trail, bridge(_by_actor(), observed_at=1))
        s2 = persist(trail, bridge(_by_actor(), observed_at=2))
        s3 = persist(trail, bridge(_by_actor(), observed_at=2))  # idempotent re-run
        assert s1["added"] > 0
        assert s2["added"] == s1["added"] and s2["duplicate"] == 0
        assert s3["added"] == 0 and s3["duplicate"] == s2["added"]
        # on-disk trail holds both snapshots
        _, obs = load_trail(trail)
        ats = sorted({o[":obs/observed-at"] for o in obs})
        assert ats == [1, 2]


def test_persist_header_marks_derived_and_gated():
    edn = emit_trail_edn(*(lambda b: (b["series"],
                                      append_obs([], b["obs"])[0]))(bridge(_by_actor(), 1)))
    assert "APPEND-ONLY" in edn and "DERIVED" in edn
    assert "G10-gated" in edn and "非終末論" in edn


# ── helper: parse EDN from a string (load_edn only takes a path) ──────────────
def load_edn_from_str(s: str):
    with tempfile.NamedTemporaryFile("w", suffix=".edn", delete=False) as f:
        f.write(s)
        p = pathlib.Path(f.name)
    try:
        return load_edn(p)
    finally:
        p.unlink()


def _run():
    fns = [v for k, v in sorted(globals().items())
           if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"persist.py: {len(fns)}/{len(fns)} tests passed")
    return True


if __name__ == "__main__":
    sys.exit(0 if _run() else 1)

File: 20-actors/mitooshi/methods/test_persist.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_persist.py.
(ns mitooshi.methods.test-persist
  "Tests for the mitooshi append-only chokepoint-intel persistence (methods/persist.clj).

  Run:  bb --classpath 20-actors 20-actors/mitooshi/methods/test_persist.clj

  The load-bearing property under test: the trail is APPEND-ONLY (非終末論). A re-run is
  idempotent; a new snapshot is additive; an existing observation is NEVER removed or
  mutated; the emitted EDN round-trips through the same reader a live ingest would use."
  (:require [mitooshi.methods.bridge :as mb]
            [mitooshi.methods.persist :as mp]
            [clojure.java.io :as io]
            [clojure.edn :as edn]
            [clojure.string :as str]
            [clojure.test :refer [deftest is run-tests]]))

(def ^:private this-file *file*)

(defn- bridge-data-dir []
  (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile
      (io/file "data" "bridge")))

(defn- by-actor []
  {"watari"    (mb/load-edn (io/file (bridge-data-dir) "watari-sample.edn"))
   "watatsuna" (mb/load-edn (io/file (bridge-data-dir) "watatsuna-sample.edn"))})

;; ── helper: parse EDN from a string (load-trail only takes a path) ────────────
(defn- load-edn-from-str [s]
  (let [tmp (java.io.File/createTempFile "persist-test-" ".edn")]
    (try
      (spit tmp s)
      (let [[series obs] (mp/load-trail tmp)]
        ;; return flat list of all records, matching Python load_edn_from_str behaviour
        (concat (vals series) obs))
      (finally
        (.delete tmp)))))

;; ── tests ─────────────────────────────────────────────────────────────────────

(deftest test-append-to-empty-trail-adds-all
  (let [b (mb/bridge (by-actor) 1)
        [merged added dup] (mp/append-obs [] (get b "obs"))]
    (is (= added (count (get b "obs"))))
    (is (= dup 0))
    (is (= (count merged) (count (get b "obs"))))))

(deftest test-reappend-same-snapshot-is-idempotent
  (let [b (mb/bridge (by-actor) 1)
        [merged _ _] (mp/append-obs [] (get b "obs"))
        [merged2 added2 dup2] (mp/append-obs merged (get b "obs"))]
    ;; nothing new, all duplicates
    (is (= added2 0))
    (is (= dup2 (count (get b "obs"))))
    ;; trail unchanged
    (is (= (count merged2) (count merged)))))

(deftest test-new-snapshot-is-additive-never-removes
  (let [b1 (mb/bridge (by-actor) 1)
        b2 (mb/bridge (by-actor) 2)
        [merged _ _] (mp/append-obs [] (get b1 "obs"))
        before-ids (set (map #(get % ":obs/id") merged))
        [merged2 added2 dup2] (mp/append-obs merged (get b2 "obs"))
        after-ids (set (map #(get % ":obs/id") merged2))]
    ;; 非終末論: never drops an obs
    (is (every? #(contains? after-ids %) before-ids))
    ;; different ts → all new
    (is (= added2 (count (get b2 "obs"))))
    (is (= dup2 0))))

(deftest test-existing-obs-values-are-not-mutated
  (let [b1 (mb/bridge (by-actor) 1)
        [merged _ _] (mp/append-obs [] (get b1 "obs"))
        snapshot (into {} (map (fn [o] [(get o ":obs/id") (get o ":obs/value")]) merged))
        ;; a later (hypothetically revised) snapshot at the same ts must NOT overwrite
        [merged2 added dup] (mp/append-obs merged (get b1 "obs"))
        after (into {} (map (fn [o] [(get o ":obs/id") (get o ":obs/value")]) merged2))]
    (is (= after snapshot))
    (is (= added 0))))

(deftest test-merge-series-is-union-first-wins
  (let [b (mb/bridge (by-actor) 1)
        merged (mp/merge-series {} (get b "series"))]
    (is (= (set (keys merged)) (set (keys (get b "series")))))
    ;; re-merging keeps the first definition (stable identity)
    (let [again (mp/merge-series merged (get b "series"))]
      (is (= again merged)))))

(deftest test-emit-round-trips-through-reader
  (let [b (mb/bridge (by-actor) 1)
        [merged-obs _ _] (mp/append-obs [] (get b "obs"))
        edn-str (mp/emit-trail-edn (get b "series") merged-obs)
        recs (load-edn-from-str edn-str)
        obs    (filter #(contains? % ":obs/id") recs)
        series (filter #(contains? % ":series/id") recs)]
    (is (= (count obs) (count merged-obs)))
    (is (= (count series) (count (get b "series"))))
    ;; values survive the round-trip
    (let [vals-after  (sort (map #(get % ":obs/value") obs))
          vals-before (sort (map #(get % ":obs/value") merged-obs))]
      (is (= vals-after vals-before)))))

(deftest test-persist-to-disk-two-snapshots-accumulate
  (let [tmp-dir (java.io.File/createTempFile "persist-test-dir-" "")
        _ (.delete tmp-dir)
        _ (.mkdirs tmp-dir)
        trail (io/file tmp-dir "chokepoint-trail.kotoba.edn")]
    (try
      (let [s1 (mp/persist trail (mb/bridge (by-actor) 1))
            s2 (mp/persist trail (mb/bridge (by-actor) 2))
            s3 (mp/persist trail (mb/bridge (by-actor) 2))]   ; idempotent re-run
        (is (> (get s1 "added") 0))
        (is (= (get s2 "added") (get s1 "added")))
        (is (= (get s2 "duplicate") 0))
        (is (= (get s3 "added") 0))
        (is (= (get s3 "duplicate") (get s2 "added")))
        ;; on-disk trail holds both snapshots
        (let [[_ obs] (mp/load-trail trail)
              ats (sort (set (map #(get % ":obs/observed-at") obs)))]
          (is (= ats [1 2]))))
      (finally
        ;; clean up temp files
        (doseq [f (reverse (file-seq tmp-dir))]
          (.delete f))))))

(deftest test-persist-header-marks-derived-and-gated
  (let [b (mb/bridge (by-actor) 1)
        [merged-obs _ _] (mp/append-obs [] (get b "obs"))
        edn-str (mp/emit-trail-edn (get b "series") merged-obs)]
    (is (str/includes? edn-str "APPEND-ONLY"))
    (is (str/includes? edn-str "DERIVED"))
    (is (str/includes? edn-str "G10-gated"))
    (is (str/includes? edn-str "非終末論"))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'mitooshi.methods.test-persist)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
